src/message.c

```diff
@@ -2623,12 +2627,20 @@ msg_puts_printf(char_u *str, int maxlen)
 
     if (*p != NUL && !(silent_mode && p_verbose == 0))
     {
-	if (maxlen > 0 && STRLEN(p) > (size_t)maxlen)
-	    p[maxlen] = 0;
```

Comment:
This line is the cause of the problem, right?
This may break the string passed from a caller. When `maxlen` is 1, 'hello' is truncated to 'h'.

How about minimize the change only in this block:

```diff
--- a/src/message.c
+++ b/src/message.c
@@ -2627,12 +2627,30 @@ msg_puts_printf(char_u *str, int maxlen)
 
     if (*p != NUL && !(silent_mode && p_verbose == 0))
     {
-	if (maxlen > 0 && STRLEN(p) > (size_t)maxlen)
-	    p[maxlen] = 0;
+	char_u *q = p;
+	char_u *tofree = NULL;
+	int len = (int)STRLEN(p);
+
+	if (maxlen > 0 && len > maxlen)
+	{
+	    q = tofree = alloc(maxlen + 1);
+	    if (q != NULL)
+		vim_strncpy(q, p, maxlen);
+	    else
+		q = p;
+	}
 	if (info_message)
-	    mch_msg((char *)p);
+	    mch_msg((char *)q);
 	else
-	    mch_errmsg((char *)p);
+	    mch_errmsg((char *)q);
+	vim_free(tofree);
+
+#ifdef FEAT_RIGHTLEFT
+	if (cmdmsg_rl)
+	    msg_col -= len;
+	else
+#endif
+	    msg_col += len;
     }
 
     msg_didout = TRUE;	    // assume that line is not empty
```

This doesn't require new functions `mch_msg_len()` and `mch_errmsg_len()`.

Reply:
When `ex_echo()` is called, from https://github.com/vim/vim/blob/master/src/eval.c#L8472-L8482, `maxlen` is the length of one character (always 1 if ascii) so `msg_puts_printf()` is called [the length of the string passed to `ex_echo()`] times at most, thus also `alloc(2)` and `free()` are repeated same times. I feel it is wasteful.

Reply:
I think we can just restore the character.